File: app/src/main/java/eu/kanade/tachiyomi/data/database/models/Category.java
package eu.kanade.tachiyomi.data.database.models;

import com.pushtorefresh.storio.sqlite.annotations.StorIOSQLiteColumn;
import com.pushtorefresh.storio.sqlite.annotations.StorIOSQLiteType;

import java.io.Serializable;

import eu.kanade.tachiyomi.data.database.tables.CategoryTable;

@StorIOSQLiteType(table = CategoryTable.TABLE)
public class Category implements Serializable {

    @StorIOSQLiteColumn(name = CategoryTable.COLUMN_ID, key = true)
    public Integer id;

    @StorIOSQLiteColumn(name = CategoryTable.COLUMN_NAME)
    public String name;

    @StorIOSQLiteColumn(name = CategoryTable.COLUMN_ORDER)
    public int order;

    @StorIOSQLiteColumn(name = CategoryTable.COLUMN_FLAGS)
    public int flags;

    public Category() {}

    public static Category create(String name) {
        Category c = new Category();
        c.name = name;
        return c;
    }

    public static Category createDefault() {
        Category c = create("Default");
        c.id = 0;
        return c;
    }

    public String getNameLower() {
        return name.toLowerCase();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Category category = (Category) o;

        return name.equals(category.name);
    }

    @Override
    public int hashCode() {
        return name.hashCode();
    }
}

File: app/src/main/java/eu/kanade/tachiyomi/data/database/models/Category.kt
package eu.kanade.tachiyomi.data.database.models

import eu.kanade.tachiyomi.ui.library.setting.DisplayModeSetting
import eu.kanade.tachiyomi.ui.library.setting.SortDirectionSetting
import eu.kanade.tachiyomi.ui.library.setting.SortModeSetting
import java.io.Serializable
import eu.kanade.domain.category.model.Category as DomainCategory

interface Category : Serializable {

    var id: Int?

    var name: String

    var order: Int

    var flags: Int

    private fun setFlags(flag: Int, mask: Int) {
        flags = flags and mask.inv() or (flag and mask)
    }

    var displayMode: Int
        get() = flags and DisplayModeSetting.MASK.toInt()
        set(mode) = setFlags(mode, DisplayModeSetting.MASK.toInt())

    var sortMode: Int
        get() = flags and SortModeSetting.MASK.toInt()
        set(mode) = setFlags(mode, SortModeSetting.MASK.toInt())

    var sortDirection: Int
        get() = flags and SortDirectionSetting.MASK.toInt()
        set(mode) = setFlags(mode, SortDirectionSetting.MASK.toInt())
}

fun Category.toDomainCategory(): DomainCategory? {
    val categoryId = id ?: return null
    return DomainCategory(
        id = categoryId.toLong(),
        name = this.name,
        order = this.order.toLong(),
        flags = this.flags.toLong(),
    )
}
